README

  This is a fork of TrustedGRUB 1.1.5


  TrustedGRUB is a modified version of GRUB Legacy (1.0) that will
extend the chain of trust in the TPM by measuring the boot loader code,
the boot settings, the loaded kernel and the initial RAM disk image
used during the boot process.


  The original source of the code can be found at:
http://sourceforge.net/projects/trustedgrub/files/TrustedGRUB-1.1.5/
  Unfortunately, the project doesn't seem to be actively maintained anymore
(last update was in 2010).


Reasons for the fork:

  - The original project is not actively maintained anymore.
  - It doesn't build with the latest autoconf/automake tools.
  - The documentation is somewhat lacking.
  - The installation wasn't producing consistent measurements.
  - Github is a nicer place for continuing development.


Use the 'build.sh' script to configure and build the project.
Use the '-v' option to enable verbose output.
Refer to the file ORIGINAL_TRUSTEDGRUB_README for additional information.
Original source code archive preserved in TrustedGRUB-1.1.5.tar.gz.


Notes about installation:

  The chain of trust is only extended by stage1 and stage2. TrustedGRUB doesn't
have stage 1.5 support for extending the chain of measurements in the PCRs.
  If we don't embed stage1.5 in the drive, stage1 will need to have embeded in it
the absolute sector of stage2. This will make the measurement of stage1 differ
at each installation. Also, grub will patch the stage1 code depending on
the BIOS version detected on the machine.

  To keep the measurements consistent, I suggest another installation aproach:
newer disk partitioning tools leave 1 MB of unused space at the front of the drive,
instead of the previous standard of 32 KB, so we can embed the entire stage2 there. 
  The script 'trustedgrub_install.sh' implements this method.


Installation instructions:
  - Copy the files from bin/ into /boot/grub/
  - Partition the disk with a tool that leaves 1 MB of empty space at the beggining of the disk.
  - Configure your boot settings in the menu.lst file.
  - Execute '/boot/grub/trustedgrub_install.sh /dev/XXX'


Original installation instructions:

  If you want to use the original setup method, be sure to delete stage1.5 first.

  Please do:
    'cp default /boot/grub'
    'make install'

  To install TrustedGRUB to your local harddisc do:
    'rm -rf /boot/grub/stage*'
    'rm -rf /boot/grub/*1_5'
    'cp default /boot/grub'
    'cp stage1/stage1 /boot/grub'
    'cp stage2/stage2 /boot/grub'
    './grub/grub --no-floppy'
  Then enter:
    'root (hdX,Y)'
    'setup (hdX)'
    'quit'
  or alternatively:
    'rm -rf /boot/grub/stage*'
    'rm -rf /boot/grub/*1_5'
    'util/grub-install /dev/XXX --no-floppy'


PCRs measurements:

  PCR 04: "Calling INT 19h", separator event, "Returned INT 19h", first 440 bytes of stage1/first sector.
  PCR 05: Separator event, last 72 bytes of stage1/first sector that contain the partition table.
  PCR 08: First 512 bytes of stage2.
  PCR 09: All of stage2 code without first 512 bytes.
  PCR 12: Executed commands from menu.lst (root, kernel, initrd, boot).
  PCR 13: 'checkfile' option checked files.
  PCR 14: Loaded files (kernel and initrd image).


Changes:

28/09/2014:
  - Renamed GRUB's 'README' file to 'ORIGINAL_GRUB_README'.
  - Renamed TrustedGRUB's 'README' file to 'ORIGINAL_TRUSTEDGRUB_README'.
  - Renamed TrustedGRUB build script from 'build_tgrub.sh' to 'build.sh'
  - Updated project to work with the latest autoconf/automake.
  - Updated build script to create the consistent measurements installation binaries.


ChangeLog
This file describes the changes between different versions of TrustedGRUB.

/* TrustedGRUB v1.1.5 */
  - Added OpenBSD and FreeBSD patches from http://sourceforge.net/projects/bsssd
  - Fixed kernel loading from command line
  - Fixed counter variable in stage2

/* TrustedGRUB v1.1.4 */

  - Ext4 support
  - Applied the following GNU GRUB patches from Gentoo repository to TrustedGRUB:
    - 008_all_grub-0.97-AM_PROG_AS.patch
    - 110_all_grub-0.97-ptable-fix.patch
    - 550_all_grub-0.97-long-commandline.patch
    - 700_all_grub-0.97-grub-install_addsyncs.patch
    - 710_all_grub-0.97-grub-install_regexp.patch
    - 720_all_grub-0.97-grub-install_aoe_support.patch
    - 800_all_grub-0.97-crossreference_manpages.patch
    - 810_all_grub-0.97-ext3_256byte_inode.patch
    - 822_all_grub-0.97-geometry-26kernel.patch
    - 850_all_grub-0.97_ext4.patch
  - Fixed compile issue on Debian-based systems
    - (GRUB requires a working absolute objcopy; upgrade your binutils)
    - (__stack_chk_fail)
  - Fixed make check
  - Hardcoded stage2 start address in stage1.S
  - Added new VESA3 support (new commands vbeset, vbeprobe, vbematch)

/* TrustedGRUB v1.1.3 */

  - Support for Linux versions 2.6.23 and above
  - Improved Checkfile-output

/* TrustedGRUB v1.1.2 */

  - GCC-4 compilation support

/* TrustedGRUB v1.1.0 */

  - fixed the measurement-process of arbitrary kernels (e.g., XEN)
  - fixed some output, when "-DSHOW_SHA1" is set
  - fixed some minors and added some options in the "build_TrustedGRUB"-script

/* TrustedGRUB v1.0.1 */

  - added cmdline-measurement into PCR12
  - fixed PCR-update for stage2_eltorito-images
  - added GCC-version verification in build_TrustedGRUB.sh

/* TrustedGRUB v1.0-rc5-ntfs */

  - added NTFS-read patch

/* TrustedGRUB v1.0-rc5 */

  - corrected PCR-update mechanism for TPM 1.2 BIOS implementations
  - "savedefault": excluded from measurement
  - beauty corrections

/* TrustedGRUB v1.0-rc4 */

  - Updated to latest GRUB Release 0.9.7
  - Including password dialog support
  - Including patches for running L4 microkernels (e.g., L4 fiasco)
  - New Software-SHA1-implementation
  - Improved checkfile option with software measurements
  - TPM Support with TPM detection
  - TPM hashing improvements
  - New "sha1"-commandline utility
  - New "verify_pcr" utility to verify the correctness of Trusted GRUB
  - New "create_sha1" utility

/* TrustedGRUB v0.8.1 */

  - corrected some minor (not security critical) bugs in the TrustedGRUB
    enhancements. Thanks to Marko Wolf for recognizing them in the source 
    code
  - the comments within TrustedGRUB's source code have been extended, especially
    in the load_checkfile function
  - the file README.TrustedGRUB has been slightly modified
  - corrected a bug in (original) GRUB which could lead to problems during
    compilation (wrong data type within the file fsys_reiserfs.c)


/* TrustedGRUB v0.8 */

  - TrustedGRUB's measurement unit has been completely redisgned. Instead of loading
    components (as the kernel) twice, it is now loaded only once, thus
    preventing a race condition which could be used for attack purposes.
    Thanks to Bernhard Kauer for presenting us the security leak which was
    present in previous version.
  - in order to reach the security improvements described above, 2 new
    functions (TrustedGRUB_open and TrustedGRUB_read) have been added to the file 
    disk_io.c


/* TrustedGRUB v0.7.1 */

  - TrustedGRUB now uses different PCRs (10 and 11) for storing hash values of
    operating system components (as, e.g., kernel, modules) and hash values
    of the check file and the files listed in it
  - added a file BUGS.TrustedGRUB


/* TrustedGRUB v0.7 */

  - instead of transferring single bytes to the TCG chip, TrustedGRUB is now able to
    uses larger areas of memory instead (more efficient)
  - MD5 hash algorithm not necessary any more; everything is hashed via the
    TCG extension
  - for easy creation of reference hash values, a program named create_sha1 has
    been added


/* TrustedGRUB v0.6 */

  - TrustedGRUB now able to check integrity of arbitrary files
  - integrity of commands now guaranteed
  - changes for TCG segment: is now 0x8000 (i.e., address 0x80000) for all
    components
  - patches included for easily transforming an original GRUB into TrustedGRUB

	
/* TrustedGRUB v0.5 */

  - fixed major bug in file asm.S (values have not been correctly measured)
  - minimal corrections in README file
	

/* TrustedGRUB v0.4.1 */

  - corrected some minor bug within file stage1.S	
  - README file updated
  - renamed HISTORY to ChangeLog (this file)
	

/* TrustedGRUB v0.4 */

  - binary versions included (both for hard and floppy drives)
  - original source code of stage1 included again (as commentary)
  - README file updated
  - HISTORY file added


/* TrustedGRUB v0.3 */

  - hard drive support added
  - README file updated


/* TrustedGRUB v0.2.1 */

  - initrd (initial ramdisk) is now also measured


/* TrustedGRUB v0.2 */

  - start.S now dynamically computes size of the rest of stage2 instead of
    using a fixed value


/* TrustedGRUB v0.1 */

  - initial version
  - full functionality (chain of trust is complete) 
  - supports floppy disks

stage2/start.S
/*
 *  GRUB  --  GRand Unified Bootloader
 *  Copyright (C) 1999,2000,2001   Free Software Foundation, Inc.
 *
 *  This program is free software; you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation; either version 2 of the License, or
 *  (at your option) any later version.
 *
 *  This program is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *  along with this program; if not, write to the Free Software
 *  Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.
 */

#define ASM_FILE
#include <shared.h>

#ifndef STAGE1_5
#include <stage2_size.h>
#endif
	
/*
 *  defines for the code go here
 */

	/* Absolute addresses
	   This makes the assembler generate the address without support
	   from the linker. (ELF can't relocate 16-bit addresses!) */
#ifdef STAGE1_5
# define ABS(x) (x-_start+0x2000)
#else
# define ABS(x) (x-_start+0x8000)
#endif /* STAGE1_5 */
	
	/* Print message string */
#define MSG(x)	movw $ABS(x), %si; call message

	.file	"start.S"

	.text

	/* Tell GAS to generate 16-bit instructions so that this code works
	   in real mode. */
	.code16

	.globl	start, _start
start:
_start:

	
	/*
	 * _start is loaded at 0x8000 and is jumped to with
	 * CS:IP 0:0x8000 in stage2.
	 */

	/* 
	 * we continue to use the stack for stage1 and assume that
	 * some registers are set to correct values. See stage1.S
	 * for more information.
	 */
	
	/* save drive reference first thing! */
	pushw	%dx

	/* print a notification message on the screen */
	pushw	%si
	MSG(notification_string)
	popw	%si
	
	/* this sets up for the first run through "bootloop" */
	movw	$ABS(firstlist - BOOTSEC_LISTSIZE), %di

	/* save the sector number of the second sector in %ebp */
	movl	(%di), %ebp

        /* this is the loop for reading the secondary boot-loader in */
bootloop:

	/* check the number of sectors to read */
	cmpw	$0, 4(%di)

	/* if zero, go to the start function */
	je	bootit

setup_sectors:	
	/* check if we use LBA or CHS */
	cmpb	$0, -1(%si)

	/* jump to chs_mode if zero */
	je	chs_mode

lba_mode:	
	/* load logical sector start */
	movl	(%di), %ebx

	/* the maximum is limited to 0x7f because of Phoenix EDD */
	xorl	%eax, %eax
	movb	$0x7f, %al

	/* how many do we really want to read? */
	cmpw	%ax, 4(%di)	/* compare against total number of sectors */

	/* which is greater? */
	jg	1f

	/* if less than, set to total */
	movw	4(%di), %ax

1:	
	/* subtract from total */
	subw	%ax, 4(%di)

	/* add into logical sector start */
	addl	%eax, (%di)

	/* set up disk address packet */

	/* the size and the reserved byte */
	movw	$0x0010, (%si)

	/* the number of sectors */
	movw	%ax, 2(%si)

	/* the absolute address (low 32 bits) */
	movl	%ebx, 8(%si)

	/* the segment of buffer address */
	movw	$BUFFERSEG, 6(%si)

	/* save %ax from destruction! */
	pushw	%ax

	/* zero %eax */
	xorl	%eax, %eax

	/* the offset of buffer address */
	movw	%ax, 4(%si)

	/* the absolute address (high 32 bits) */
	movl	%eax, 12(%si)


/*
 * BIOS call "INT 0x13 Function 0x42" to read sectors from disk into memory
 *	Call with	%ah = 0x42
 *			%dl = drive number
 *			%ds:%si = segment:offset of disk address packet
 *	Return:
 *			%al = 0x0 on success; err code on failure
 */

	movb	$0x42, %ah
	int	$0x13

	jc	read_error

	movw	$BUFFERSEG, %bx
	jmp	copy_buffer
			
chs_mode:	
	/* load logical sector start (bottom half) */
	movl	(%di), %eax

	/* zero %edx */
	xorl	%edx, %edx

	/* divide by number of sectors */
	divl	(%si)

	/* save sector start */
	movb	%dl, 10(%si)

	xorl	%edx, %edx	/* zero %edx */
	divl	4(%si)		/* divide by number of heads */

	/* save head start */
	movb	%dl, 11(%si)

	/* save cylinder start */
	movw	%ax, 12(%si)

	/* do we need too many cylinders? */
	cmpw	8(%si), %ax
	jge	geometry_error

	/* determine the maximum sector length of this read */
	movw	(%si), %ax	/* get number of sectors per track/head */

	/* subtract sector start */
	subb	10(%si), %al

	/* how many do we really want to read? */
	cmpw	%ax, 4(%di)	/* compare against total number of sectors */


	/* which is greater? */
	jg	2f

	/* if less than, set to total */
	movw	4(%di), %ax

2:	
	/* subtract from total */
	subw	%ax, 4(%di)

	/* add into logical sector start */
	addl	%eax, (%di)

/*
 *  This is the loop for taking care of BIOS geometry translation (ugh!)
 */

	/* get high bits of cylinder */
	movb	13(%si), %dl

	shlb	$6, %dl		/* shift left by 6 bits */
	movb	10(%si), %cl	/* get sector */

	incb	%cl		/* normalize sector (sectors go
					from 1-N, not 0-(N-1) ) */
	orb	%dl, %cl	/* composite together */
	movb	12(%si), %ch	/* sector+hcyl in cl, cylinder in ch */

	/* restore %dx */
	popw	%dx
	pushw	%dx

	/* head number */
	movb	11(%si), %dh

	pushw	%ax	/* save %ax from destruction! */

/*
 * BIOS call "INT 0x13 Function 0x2" to read sectors from disk into memory
 *	Call with	%ah = 0x2
 *			%al = number of sectors
 *			%ch = cylinder
 *			%cl = sector (bits 6-7 are high bits of "cylinder")
 *			%dh = head
 *			%dl = drive (0x80 for hard disk, 0x0 for floppy disk)
 *			%es:%bx = segment:offset of buffer
 *	Return:
 *			%al = 0x0 on success; err code on failure
 */

	movw	$BUFFERSEG, %bx
	movw	%bx, %es	/* load %es segment with disk buffer */

	xorw	%bx, %bx	/* %bx = 0, put it at 0 in the segment */
	movb	$0x2, %ah	/* function 2 */
	int	$0x13

	jc	read_error

	/* save source segment */
	movw	%es, %bx
	
copy_buffer:	

	/* load addresses for copy from disk buffer to destination */
	movw	6(%di), %es	/* load destination segment */

	/* restore %ax */
	popw	%ax

	/* determine the next possible destination address (presuming
		512 byte sectors!) */
	shlw	$5, %ax		/* shift %ax five bits to the left */
	addw	%ax, 6(%di)	/* add the corrected value to the destination
				   address for next time */

	/* save addressing regs */
	pusha
	pushw	%ds

	/* get the copy length */
	shlw	$4, %ax
	movw	%ax, %cx

	xorw	%di, %di	/* zero offset of destination addresses */
	xorw	%si, %si	/* zero offset of source addresses */
	movw	%bx, %ds	/* restore the source segment */

	cld		/* sets the copy direction to forward */

	/* perform copy */
	rep		/* sets a repeat */
	movsb		/* this runs the actual copy */

	/* restore addressing regs and print a dot with correct DS 
	   (MSG modifies SI, which is saved, and unused AX and BX) */
	popw	%ds
	MSG(notification_step)
	popa

	/* check if finished with this dataset */
	cmpw	$0, 4(%di)
	jne	setup_sectors

	/* update position to load from */
	subw	$BOOTSEC_LISTSIZE, %di

	/* jump to bootloop */
	jmp	bootloop

/* END OF MAIN LOOP */

bootit:
	/* print a newline */
	MSG(notification_done)
	popw	%dx	/* this makes sure %dl is our "boot" drive */
#ifdef STAGE1_5
	ljmp	$0, $0x2200
#else /* ! STAGE1_5 */

	
	/* Begin TCG extension */
	/* Hashes the rest of stage2 and writes the result into PCR9.
	   For details see README file. */
	
	/* Store registers which have to be modified */
	pushw %es
	pushw %ds
	pushw %si
	pushw %di
	pushl %edx

	/* Create own segment for TCG */
	movw $TCG_SEG, %ax
	movw %ax, %es
	movw %ax, %ds
	
	/* Preparing and executing TCG_HashAll function call */
	movw $0x0010, 0x00		/* length input parameter block */
	movl $0x00008200, 0x04		/* start address of stage2 (part2) to be hashed */
	movl $(STAGE2_SIZE), 0x08	/* length of data to be hashed */
	movl $0x00000004, 0x0C		/* algorithm = SHA1 */

	movw $0xBB05, %ax		/* function = TCG_HashAll */
	movl $0x41504354, %ebx		/* TCG value */
	xorl %ecx, %ecx			/* must be zero */
	xorl %edx, %edx			/* must be zero */
	movw $0x00, %di			/* pointer to input block */
	movw $0x28, %si			/* pointer to digest block */
	int $0x1A				/* TCG interrupt call */

	/* Preparing and executing TCG_PassThroughToTPM function call */
	movw $0x002A, 0x12		/* length input parameter block */
	movw $0x0022, 0x16		/* size of output parameter block */
	movw $0xC100, 0x1A		/* TPM_TAG_RQU_COMMAND */
	movl $0x22000000, 0x1C		/* total number of input bytes */
	movl $0x14000000, 0x20		/* command ordinal */
	movl $0x09000000, 0x24		/* PCR to be updated */
	
	movw $0xBB02, %ax		/* function = TCG_PassThroughToTPM */
	movl $0x41504354, %ebx		/* TCG value */
	movw $0x12, %di			/* pointer to input block */
	movw $0x60, %si			/* pointer to output block */
	int $0x1A			/* TCG interrupt call */

	/* Restore registers saved at the beginning */
	popl %edx
	popw %di
	popw %si
	popw %ds
	popw %es
		
	/* End TCG extension */
	
	
	ljmp	$0, $0x8200
#endif /* ! STAGE1_5 */


/*
 * BIOS Geometry translation error (past the end of the disk geometry!).
 */
geometry_error:
	MSG(geometry_error_string)
	jmp	general_error

/*
 * Read error on the disk.
 */
read_error:
	MSG(read_error_string)

general_error:
	MSG(general_error_string)

/* go here when you need to stop the machine hard after an error condition */
stop:	jmp	stop

#ifdef STAGE1_5
notification_string:	.string "Loading stage1.5"
#else
notification_string:	.string "Loading stage2"
#endif

notification_step:	.string "."
notification_done:	.string "\r\n"
	
geometry_error_string:	.string "Geom"
read_error_string:	.string "Read"
general_error_string:	.string " Error"

/*
 * message: write the string pointed to by %si
 *
 *   WARNING: trashes %si, %ax, and %bx
 */

	/*
	 * Use BIOS "int 10H Function 0Eh" to write character in teletype mode
	 *	%ah = 0xe	%al = character
	 *	%bh = page	%bl = foreground color (graphics modes)
	 */
1:
	movw	$0x0001, %bx
	movb	$0xe, %ah
	int	$0x10		/* display a byte */

	incw	%si
message:
	movb	(%si), %al
	cmpb	$0, %al
	jne	1b	/* if not end of string, jmp to display */
	ret
lastlist:

/*
 *  This area is an empty space between the main body of code below which
 *  grows up (fixed after compilation, but between releases it may change
 *  in size easily), and the lists of sectors to read, which grows down
 *  from a fixed top location.
 */

	.word 0
	.word 0

	. = _start + 0x200 - BOOTSEC_LISTSIZE
	
        /* fill the first data listing with the default */
blocklist_default_start:
	.long 2		/* this is the sector start parameter, in logical
			   sectors from the start of the disk, sector 0 */
blocklist_default_len:
			/* this is the number of sectors to read */
#ifdef STAGE1_5
	.word 0		/* the command "install" will fill this up */
#else
	.word (STAGE2_SIZE + 511) >> 9
#endif
blocklist_default_seg:
#ifdef STAGE1_5
	.word 0x220
#else
	.word 0x820	/* this is the segment of the starting address
			   to load the data into */
#endif
	
firstlist:	/* this label has to be after the list data!!! */
